core/engine.go

```diff
@@ -89,89 +84,32 @@ func NewEngine(
 
 	e := &Engine{
 		ExecutionScheduler: ex,
-		executionState:     ex.GetState(),
 
-		options:        opts,
 		runtimeOptions: rtOpts,
-		outputs:        outputs,
-		Metrics:        make(map[string]*stats.Metric),
 		Samples:        make(chan stats.SampleContainer, opts.MetricSamplesBufferSize.Int64),
 		stopChan:       make(chan struct{}),
 		logger:         logger.WithField("component", "engine"),
-		registry:       registry,
-	}
-
-	if !(e.runtimeOptions.NoSummary.Bool && e.runtimeOptions.NoThresholds.Bool) {
-		err := e.initSubMetricsAndThresholds()
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	return e, nil
-}
-
-func (e *Engine) getOrInitPotentialSubmetric(name string) (*stats.Metric, error) {
-	// TODO: replace with strings.Cut after Go 1.18
-	nameParts := strings.SplitN(name, "{", 2)
-
-	metric := e.registry.Get(nameParts[0])
-	if metric == nil {
-		return nil, fmt.Errorf("metric '%s' does not exist in the script", nameParts[0])
-	}
-	if len(nameParts) == 1 { // no sub-metric
-		return metric, nil
 	}
 
-	if nameParts[1][len(nameParts[1])-1] != '}' {
-		return nil, fmt.Errorf("missing ending bracket, sub-metric format needs to be 'metric{key:value}'")
-	}
-	sm, err := metric.AddSubmetric(nameParts[1][:len(nameParts[1])-1])
+	me, err := engine.NewMetricsEngine(registry, ex.GetState(), opts, rtOpts, logger)
 	if err != nil {
 		return nil, err
 	}
-	return sm.Metric, nil
-}
+	e.MetricsEngine = me
 
-func (e *Engine) initSubMetricsAndThresholds() error {
-	for metricName, thresholds := range e.options.Thresholds {
-		metric, err := e.getOrInitPotentialSubmetric(metricName)
-
-		if e.runtimeOptions.NoThresholds.Bool {
-			if err != nil {
-				e.logger.WithError(err).Warnf("Invalid metric '%s' in threshold definitions", metricName)
-			}
-			continue
-		}
-
-		if err != nil {
-			return fmt.Errorf("invalid metric '%s' in threshold definitions: %w", metricName, err)
-		}
-
-		metric.Thresholds = thresholds
-		e.metricsWithThresholds = append(e.metricsWithThresholds, metric)
-
-		// Mark the metric (and the parent metricq, if we're dealing with a
-		// submetric) as observed, so they are shown in the end-of-test summary,
-		// even if they don't have any metric samples during the test run
-		metric.Observed = true
-		e.Metrics[metric.Name] = metric
-		if metric.Sub != nil {
-			metric.Sub.Metric.Observed = true
-			e.Metrics[metric.Sub.Metric.Name] = metric.Sub.Metric
-		}
+	if !(rtOpts.NoSummary.Bool && rtOpts.NoThresholds.Bool) {
+		e.ingester = me.GetIngester()
+		outputs = append(outputs, e.ingester)
```

Comment:
Is it not possible to have this in the same place where we set the other outputs?

Reply:
It may be, but I think it will actually be worse to do it that way. 

he [other outputs receive a whole bunch of things](https://github.com/grafana/k6/blob/f06d6f0b673ebc9884995478a4f8d5d7ac03a923/cmd/outputs.go#L80-L130) this one doesn't need, while this one needs the `MetricsEngine`, something we create a few lines above here (or in `cmd/run.go` in the follow up PR without an Engine). Besides, we add this output based on completely different criteria only some of the time, based on RuntimeOptions values...